src/main/software/amazon/event/ruler/SubRuleContext.java

```diff
@@ -11,63 +11,52 @@
  * A sub-rule refers to name/value pairs, usually represented by Map of String to List of Patterns, that compose a rule.
  * In the case of $or, one rule will have multiple name/value pairs, and this is why we use the "sub-rule" terminology.
  */
-public class SubRuleContext {
+public final class SubRuleContext {
 
-    private final double id;
+    private final long id;
+    private final Object ruleName;
 
-    private SubRuleContext(double id) {
+    SubRuleContext(long id, Object ruleName) {
         this.id = id;
+        this.ruleName = ruleName;
     }
 
-    public double getId() {
-        return id;
+    public Object getRuleName() {
+        return ruleName;
     }
 
     @Override
-    public boolean equals(Object o) {
-        if (o == null || !(o instanceof SubRuleContext)) {
-            return false;
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
         }
-        SubRuleContext otherContext = (SubRuleContext) o;
-        return id == otherContext.id;
+        if (obj instanceof SubRuleContext) {
+            return id == ((SubRuleContext) obj).id;
+        }
+        return false;
     }
 
     @Override
     public int hashCode() {
-        return Double.hashCode(id);
+        return Long.hashCode(id);
     }
 
     /**
      * Generator for SubRuleContexts.
      */
     static final class Generator {
 
-        private double nextId = -Double.MAX_VALUE;
-
-        private final Map<Object, Set<Double>> nameToIds = new ConcurrentHashMap<>();
-        private final Map<Double, Object> idToName = new ConcurrentHashMap<>();
+        private final Map<Object, Set<SubRuleContext>> nameToContext = new ConcurrentHashMap<>();
+        private long nextId;
```

Comment:
Should this be AtomicLong ?

Reply:
From what I see the only place where it is called is `GenericMachine.addPatternRule()` under `synchronized(this)`, so `AtomicLong` does not make much sense.